obspy/clients/fdsn/tests/test_client.py

```diff
@@ -1448,15 +1456,96 @@ def test_get_waveforms_empty_seed_codes(self):
             url_parts = url.replace(url_base, '').split("&")
             self.assertIn('{}='.format(key), url_parts)
 
-    def test_no_data(self):
+    @mock.patch("obspy.clients.fdsn.client.download_url")
+    def test_no_data_exception(self, download_url_mock):
         """
         Verify that a request returning no data raises an identifiable
         exception
         """
-        self.assertRaises(FDSNNoDataException, self.client.get_events,
-                          starttime=UTCDateTime("2001-01-07T01:00:00"),
-                          endtime=UTCDateTime("2001-01-07T01:01:00"),
-                          minmagnitude=8)
+        download_url_mock.return_value = (204, None)
+        self.assertRaises(FDSNNoDataException, self.client.get_stations)
+
+    @mock.patch("obspy.clients.fdsn.client.download_url")
+    def test_request_too_large_exception(self, download_url_mock):
+        """
+        Verify that a request returning too much data raises an identifiable
+        exception
+        """
+        download_url_mock.return_value = (413, None)
+        self.assertRaises(FDSNRequestTooLargeException,
+                          self.client.get_stations)
+
+    @mock.patch("obspy.clients.fdsn.client.download_url")
+    def test_authentication_exception(self, download_url_mock):
+        """
+        Verify that a request with missing authentication raises an
+        identifiable exception
+        """
+        with mock.patch("obspy.clients.fdsn.client.Client._has_eida_auth",
+                        new_callable=mock.PropertyMock,
+                        return_value=False):
+            self.assertRaises(FDSNAuthenticationException, Client,
+                              eida_token="TEST")
+
+        self.assertRaises(FDSNAuthenticationException, Client, "IRIS",
+                          eida_token="TEST", user="TEST")
+
+        download_url_mock.return_value = (400, io.BytesIO(b"token is expired"))
+        self.assertRaises(FDSNAuthenticationException,
+                          self.client.get_stations)
```

Comment:
Even though this is indeed an authentication related problem the return code is still 400 and I think it would be a bit cleaner if every return code is mapped to a unique exception.